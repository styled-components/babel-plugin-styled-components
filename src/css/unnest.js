import { prependSelector } from './utils/unnestUtils'

const unnest = node => {
  const rules = []

  // Recursively call unnest rule and return block without them
  const unnestBlock = (blockNode, selectorPath) => {
    for (const rule of blockNode.rules) {
      unnestRule(rule, selectorPath)
    }

    return Object.assign({}, blockNode, {
      rules: []
    })
  }

  // Unnest deep block and selector on rule
  const unnestRule = (ruleNode, selectorPath) => {
    const selector = (
      selectorPath ?
        prependSelector(ruleNode.selector, selectorPath) :
        ruleNode.selector
    )

    const rule = Object.assign({}, ruleNode, {
      selector
    })

    // Push first then unnest to preserve sorting
    rules.push(rule)

    // Unnest deep block and set it on new rule
    const block = unnestBlock(ruleNode.block, selector.value)
    rule.block = block
  }

  if (node.type === 'Rule') {
    unnestRule(node)
  } else if (node.type === 'Block') {
    const rule = {
      type: 'Rule',
      loc: node.loc,
      selector: {
        type: 'SelectorLiteral',
        loc: [0, 0],
        interpolations: [],
        value: '&'
      }
    }

    // See above: Push first then unnest to preserve sorting
    rules.push(rule)

    // See above: Unnest deep block and set it on new rule
    const block = unnestBlock(node)
    rule.block = block
  }

  // Return resulting rules array
  return rules
}

export default unnest
