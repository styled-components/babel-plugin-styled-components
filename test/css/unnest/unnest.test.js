import unnest from '../../../src/css/unnest';
import parse from '../../../src/css/parser';

describe('unnest', () => {
  it('should work on nested rules', () => {
    const ast = parse`
      color: red;
      background: blue;

      & > div {
        color: green;
      }
    `

    const unnested = unnest(ast)
    expect(unnested).toMatchSnapshot()
  })

  it('should work on nested rules containing pseudo selectors', () => {
    const ast = parse`
      color: red;
      background: blue;

      &:hover {
        color: green;
      }
    `

    const unnested = unnest(ast)
    expect(unnested).toMatchSnapshot()
  })
})
