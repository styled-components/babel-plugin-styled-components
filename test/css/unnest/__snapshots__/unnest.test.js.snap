exports[`unnest should work on nested rules 1`] = `
Array [
  Object {
    "block": Object {
      "declarations": Array [
        Object {
          "loc": Array [
            0,
            17,
          ],
          "property": Object {
            "loc": Array [
              0,
              12,
            ],
            "type": "PropertyLiteral",
            "value": "
      color",
          },
          "type": "Declaration",
          "value": Object {
            "interpolations": Array [],
            "loc": Array [
              13,
              17,
            ],
            "type": "ValueLiteral",
            "value": " red",
          },
        },
        Object {
          "loc": Array [
            18,
            41,
          ],
          "property": Object {
            "loc": Array [
              18,
              35,
            ],
            "type": "PropertyLiteral",
            "value": "
      background",
          },
          "type": "Declaration",
          "value": Object {
            "interpolations": Array [],
            "loc": Array [
              36,
              41,
            ],
            "type": "ValueLiteral",
            "value": " blue",
          },
        },
      ],
      "loc": Array [
        0,
        93,
      ],
      "rules": Array [],
      "type": "Block",
    },
    "loc": Array [
      0,
      93,
    ],
    "selector": Object {
      "interpolations": Array [],
      "loc": Array [
        0,
        0,
      ],
      "type": "SelectorLiteral",
      "value": "&",
    },
    "type": "Rule",
  },
  Object {
    "block": Object {
      "declarations": Array [
        Object {
          "loc": Array [
            59,
            80,
          ],
          "property": Object {
            "loc": Array [
              59,
              73,
            ],
            "type": "PropertyLiteral",
            "value": "
        color",
          },
          "type": "Declaration",
          "value": Object {
            "interpolations": Array [],
            "loc": Array [
              74,
              80,
            ],
            "type": "ValueLiteral",
            "value": " green",
          },
        },
      ],
      "loc": Array [
        59,
        88,
      ],
      "rules": Array [],
      "type": "Block",
    },
    "loc": Array [
      42,
      89,
    ],
    "selector": Object {
      "interpolations": Array [],
      "loc": Array [
        42,
        57,
      ],
      "type": "SelectorLiteral",
      "value": "

      & > div",
    },
    "type": "Rule",
  },
]
`;

exports[`unnest should work on nested rules containing pseudo selectors 1`] = `
Array [
  Object {
    "block": Object {
      "declarations": Array [
        Object {
          "loc": Array [
            0,
            17,
          ],
          "property": Object {
            "loc": Array [
              0,
              12,
            ],
            "type": "PropertyLiteral",
            "value": "
      color",
          },
          "type": "Declaration",
          "value": Object {
            "interpolations": Array [],
            "loc": Array [
              13,
              17,
            ],
            "type": "ValueLiteral",
            "value": " red",
          },
        },
        Object {
          "loc": Array [
            18,
            41,
          ],
          "property": Object {
            "loc": Array [
              18,
              35,
            ],
            "type": "PropertyLiteral",
            "value": "
      background",
          },
          "type": "Declaration",
          "value": Object {
            "interpolations": Array [],
            "loc": Array [
              36,
              41,
            ],
            "type": "ValueLiteral",
            "value": " blue",
          },
        },
      ],
      "loc": Array [
        0,
        93,
      ],
      "rules": Array [],
      "type": "Block",
    },
    "loc": Array [
      0,
      93,
    ],
    "selector": Object {
      "interpolations": Array [],
      "loc": Array [
        0,
        0,
      ],
      "type": "SelectorLiteral",
      "value": "&",
    },
    "type": "Rule",
  },
  Object {
    "block": Object {
      "declarations": Array [
        Object {
          "loc": Array [
            59,
            80,
          ],
          "property": Object {
            "loc": Array [
              59,
              73,
            ],
            "type": "PropertyLiteral",
            "value": "
        color",
          },
          "type": "Declaration",
          "value": Object {
            "interpolations": Array [],
            "loc": Array [
              74,
              80,
            ],
            "type": "ValueLiteral",
            "value": " green",
          },
        },
      ],
      "loc": Array [
        59,
        88,
      ],
      "rules": Array [],
      "type": "Block",
    },
    "loc": Array [
      42,
      89,
    ],
    "selector": Object {
      "interpolations": Array [],
      "loc": Array [
        42,
        57,
      ],
      "type": "SelectorLiteral",
      "value": "

      &:hover",
    },
    "type": "Rule",
  },
]
`;
